hibernate-core/src/main/java/org/hibernate/dialect/function/DerbyConcatFunction.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later.
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.dialect.function;
+
+import java.util.List;
+
+import org.hibernate.query.sqm.function.AbstractSqmSelfRenderingFunctionDescriptor;
+import org.hibernate.query.sqm.produce.function.StandardArgumentsValidators;
+import org.hibernate.query.sqm.produce.function.StandardFunctionReturnTypeResolvers;
+import org.hibernate.query.sqm.sql.internal.SqmParameterInterpretation;
+import org.hibernate.sql.ast.SqlAstNodeRenderingMode;
+import org.hibernate.sql.ast.SqlAstTranslator;
+import org.hibernate.sql.ast.spi.SqlAppender;
+import org.hibernate.sql.ast.tree.SqlAstNode;
+import org.hibernate.sql.ast.tree.expression.JdbcParameter;
+import org.hibernate.type.StandardBasicTypes;
+
+/**
+ * A specialized concat() function definition in which:<ol>
+ *     <li>we translate to use the concat operator ('||')</li>
+ *     <li>wrap dynamic parameters in CASTs to VARCHAR</li>
+ * </ol>
+ * <p/>
+ * This last spec is to deal with a limitation on DB2 and variants (e.g. Derby)
+ * where dynamic parameters cannot be used in concatenation unless they are being
+ * concatenated with at least one non-dynamic operand.  And even then, the rules
+ * are so convoluted as to what is allowed and when the CAST is needed and when
+ * it is not that we just go ahead and do the CASTing.
+ *
+ * @author Steve Ebersole
+ * @author Nathan Xu
+ */
+public class DerbyConcatFunction extends AbstractSqmSelfRenderingFunctionDescriptor {
```

Comment:
Could you please explain which cases need this function? I would rather see you register the concat function like this: `CommonFunctionFactory.concat_pipeOperator( queryEngine, SqlAstNodeRenderingMode.NO_PLAIN_PARAMETER )`

Reply:
It would fail in Derby CI mode for the following testing case (org.hibernate.orm.test.jpa.criteria.basic.ConcatTest:
```
@Test
	public void testConcat() throws Exception {
		EntityManager entityManager = getOrCreateEntityManager();
		entityManager.getTransaction().begin();
		try {
			CriteriaBuilder cb = entityManager.getCriteriaBuilder();
			CriteriaQuery query = cb.createQuery();
			Root<TestEntity> testEntity = query.from( TestEntity.class );

			query.select( testEntity ).where( cb.equal( testEntity.get( "name" ), cb.concat( "test", cb.literal( "_1" ) ) ) );

			final List results = entityManager.createQuery( query ).getResultList();
			entityManager.getTransaction().commit();

			assertThat( results.size(), is( 1 ) );
		}
		catch (Exception e) {
			if ( entityManager.getTransaction().isActive() ) {
				entityManager.getTransaction().rollback();
			}
			throw e;
		}
		finally {
			entityManager.close();
		}
	}
```

The reason is Derby needs special treatment when jdbc parameter is involved, as in the above testing case. See DerbyConcatFunction in v5 for further details.

Reply:
Seems this function is not common and tightly coupled with Derby dialect.

Reply:
I tried you suggestion on org.hibernate.orm.test.jpa.criteria.basic.ConcatTest and found there is a stackoverflow issue as the following code would trigger infinite loop:
```
<img width="1016" alt="Screen Shot 2021-06-04 at 11 52 43" src="https://user-images.githubusercontent.com/8211458/120829415-61341180-c52b-11eb-9ce7-77c89f3d4864.png">
```
due to the internal implementation to append empty string ending up with infinite invoking loops.

Reply:
I added a new testing case to cover this troublesome Derby concat scenario. I think the issue still remains when we use jdbc parameter explicitly as opposed to constant value or literal as in the above testConcat() method.